sqlmesh/core/node.py

```diff
@@ -0,0 +1,169 @@
+from __future__ import annotations
+
+import typing as t
+from enum import Enum
+
+from pydantic import validator
+from sqlglot import exp
+
+from sqlmesh.utils.cron import CroniterCache
+from sqlmesh.utils.date import TimeLike, to_datetime
+from sqlmesh.utils.errors import ConfigError
+from sqlmesh.utils.pydantic import PydanticModel
+
+
+class IntervalUnit(str, Enum):
+    """IntervalUnit is the inferred granularity of an incremental model.
+
+    IntervalUnit can be one of 5 types, YEAR, MONTH, DAY, HOUR, MINUTE. The unit is inferred
+    based on the cron schedule of a model. The minimum time delta between a sample set of dates
+    is used to determine which unit a model's schedule is.
+    """
+
+    YEAR = "year"
+    MONTH = "month"
+    DAY = "day"
+    HOUR = "hour"
+    MINUTE = "minute"
+
+    @property
+    def is_date_granularity(self) -> bool:
+        return self in (IntervalUnit.YEAR, IntervalUnit.MONTH, IntervalUnit.DAY)
+
+
+class Node(PydanticModel):
+    """
+    Args:
+        start: The earliest date that the model will be backfilled for. If this is None,
+            then the date is inferred by taking the most recent start date of its ancestors.
+            The start date can be a static datetime or a relative datetime like "1 year ago"
+    """
+
+    start: t.Optional[TimeLike]
+    cron: str = "@daily"
+
+    _croniter: t.Optional[CroniterCache] = None
+    _interval_unit: t.Optional[IntervalUnit] = None
+
+    @validator("start", pre=True)
+    def _date_validator(cls, v: t.Any) -> t.Optional[TimeLike]:
+        if isinstance(v, exp.Expression):
+            v = v.name
+        if v and not to_datetime(v):
+            raise ConfigError(f"'{v}' needs to be time-like: https://pypi.org/project/dateparser")
+        return v
+
+    @validator("cron", pre=True)
+    def _cron_validator(cls, v: t.Any) -> t.Optional[str]:
+        cron = str_or_exp_to_str(v)
+        if cron:
+            from croniter import CroniterBadCronError, croniter
+
+            try:
+                croniter(cron)
+            except CroniterBadCronError:
+                raise ConfigError(f"Invalid cron expression '{cron}'")
+        return cron
+
+    @property
+    def batch_size(self) -> t.Optional[int]:
```

Comment:
what do you envision this for?

Reply:
The reason I added batch_size to node is so the scheduler and engine adapters wouldn't need model specific logic to execute nodes. I envision it as a no-op (default value of None) for non-model nodes. Model will override to model specific configuration.